File: JAP_SSH_LOCAL_PYTHON/JAP/TUNNEL.py
"""
JAP
Copyright (C) 2012 Jeroen Van Steirteghem

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
"""

from twisted.protocols import basic
from twisted.internet import protocol, reactor
import base64

class TunnelProtocol(basic.LineReceiver):
    delimiter = "\r\n"

    def __init__(self):
        self.headerLines = []
        
    def connectionMade(self):
        request = "CONNECT " + str(self.factory.address) + ":" + str(self.factory.port) + " HTTP/1.0\r\n"
        
        if self.factory.proxyServerUsername != "":
            request = request + "Proxy-Authorization: Basic " + base64.standard_b64encode(self.factory.proxyServerUsername + ":" + self.factory.proxyServerPassword) + "\r\n"
        
        request = request + "\r\n"
        
        self.transport.write(request)
        
    def lineReceived(self, line):
        if line:
            self.headerLines.append(line)
        else:
            self.setRawMode()
            v, statusCode, s = self.headerLines[0].split(" ", 2)
            lineBuffer = self.clearLineBuffer()
            if statusCode != "200":
                self.transport.loseConnection()
                return
            peer = self.transport.getPeer()
            otherProtocol = self.factory.otherProtocolFactory.buildProtocol(peer)
            self.transport.protocol = otherProtocol
            otherProtocol.transport = self.transport
            self.transport = None
            if self.factory.contextFactory:
                otherProtocol.transport.startTLS(self.factory.contextFactory)
            otherProtocol.connectionMade()
            if lineBuffer:
                otherProtocol.dataReceived(lineBuffer)


class TunnelFactory(protocol.ClientFactory):
    protocol = TunnelProtocol

    def __init__(self, address, port, proxyServerAddress, proxyServerPort, proxyServerUsername, proxyServerPassword, otherProtocolFactory, contextFactory=None):
        self.address = address
        self.port = port
        self.proxyServerAddress = proxyServerAddress
        self.proxyServerPort = proxyServerPort
        self.proxyServerUsername = proxyServerUsername
        self.proxyServerPassword = proxyServerPassword
        self.otherProtocolFactory = otherProtocolFactory
        self.contextFactory = contextFactory


class Tunnel(object):
    def __init__(self, proxyServerAddress, proxyServerPort, proxyServerUsername, proxyServerPassword):
        self.proxyServerAddress = proxyServerAddress
        self.proxyServerPort = proxyServerPort
        self.proxyServerUsername = proxyServerUsername
        self.proxyServerPassword = proxyServerPassword

    def connectTCP(self, address, port, factory, *args, **kwargs):
        tunnel_factory = TunnelFactory(address, port, self.proxyServerAddress, self.proxyServerPort, self.proxyServerUsername, self.proxyServerPassword, factory)
        return reactor.connectTCP(self.proxyServerAddress, self.proxyServerPort, tunnel_factory, *args, **kwargs)

    def connectSSL(self, address, port, factory, contextFactory, *args, **kwargs):
        tunnel_factory = TunnelFactory(address, port, self.proxyServerAddress, self.proxyServerPort, self.proxyServerUsername, self.proxyServerPassword, factory, contextFactory)
        return reactor.connectTCP(self.proxyServerAddress, self.proxyServerPort, tunnel_factory, *args, **kwargs)

File: JAP_WS_LOCAL_NODE/JAP/TUNNEL.js
/*
JAP
Copyright (C) 2012 Jeroen Van Steirteghem

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

var util = require("util");
var events = require("events");
var http = require("http");
var tls = require("tls");

function HTTPAgent(options) {
	events.EventEmitter.call(this);
	
	var self = this;
	self.options = options || {};
	
	self.createConnection = function(options, cb) {
		var requestOptions = {};
		requestOptions.host = options.proxy.host;
		requestOptions.port = options.proxy.port;
		requestOptions.method = "CONNECT";
		requestOptions.path = options.host + ":" + options.port;
		requestOptions.headers = {};
		requestOptions.headers["Proxy-Authorization"] = "Basic " + new Buffer(options.proxy.auth).toString("base64");
		requestOptions.agent = http.globalAgent;
		
		var request = http.request(requestOptions);
		
		request.on("connect", function(response, socket, head) {
			cb(socket);
		});
		
		request.on("error", function(error) {
			options.request.emit("error", error);
		});
		
		request.end();
	}
}

util.inherits(HTTPAgent, events.EventEmitter);

HTTPAgent.prototype.addRequest = function(request, host, port, localAddress) {
	var self = this;
	
	var options = util._extend({}, self.options);
	options.request = request;
	options.host = host;
	options.port = port;
	options.localAddress = localAddress;
	
	self.createSocket(options, function(socket) {
		request.onSocket(socket);
	});
};

HTTPAgent.prototype.createSocket = function(options, cb) {
	var self = this;
	
	self.createConnection(options, function(socket) {
		cb(socket);
	});
};

function HTTPSAgent(options)
{
	HTTPAgent.call(this, options);
};

util.inherits(HTTPSAgent, HTTPAgent);

HTTPSAgent.prototype.createSocket = function(options, cb) {
	var self = this;
	
	self.createConnection(options, function(socket) {
		options.socket = socket;
		socket = tls.connect(options);
		
		cb(socket);
	});
};

module.exports.HTTPAgent = HTTPAgent;
module.exports.HTTPSAgent = HTTPSAgent;